### PR TITLE
Only run restore callbacks if in window and not frozen

### PR DESCRIPTION
**Pull request by Billiam**
https://github.com/redguava/Cliniko/issues/12558

### CODE REVIEW

**Comment by Billiam** [comment]
I moved the callback block around here, so that it will only run before and after callbacks if the record is within the recovery window.

I _also_ changed the call to restore associated records inside of the same check. I can't think of why we'd want to restore related records for a given record if it failed the recovery window check (or the check for frozen-ness).

File: lib/paranoia.rb [hunk 1]
```diff
@@ -102,12 +102,12 @@ def paranoia_delete
...
-      run_callbacks(:restore) do
-        recovery_window_range = get_recovery_window_range(opts)
-        # Fixes a bug where the build would error because attributes were frozen.
-        # This only happened on Rails versions earlier than 4.1.
-        noop_if_frozen = ActiveRecord.version < Gem::Version.new("4.1")
-        if within_recovery_window?(recovery_window_range) && ((noop_if_frozen && !@attributes.frozen?) || !noop_if_frozen)
+      recovery_window_range = get_recovery_window_range(opts)
+      # Fixes a bug where the build would error because attributes were frozen.
+      # This only happened on Rails versions earlier than 4.1.
+      noop_if_frozen = ActiveRecord.version < Gem::Version.new("4.1")
```
Comment (bpinto) [inline lib/paranoia.rb:108]: Seems we could remove this? This will never be true anymore.

Comment (Billiam) [reply lib/paranoia.rb:108]: I don't think we need it, but it's not hurting us. I'd rather make rebasing with upstream easier if it's not going to change anything for us

**Review by mphalliday — APPROVED** [review]
Specs I wrote here https://github.com/redguava/Cliniko/pull/12579 pass 🎉 ... looks good to me @Billiam

**Pull request merged**